R/Rules-methods.R

```diff
@@ -717,6 +717,119 @@ setMethod(
   }
 )
 
+## NextBestTDsamples ----
+
+#' @describeIn nextBest find the next best dose based only on the DLE responses
+#'   and for [`LogisticIndepBeta`] model class object involving DLE samples.
+#'
+#' @aliases nextBest-NextBestTDsamples
+#'
+#' @export
+#' @example examples/Rules-method-nextBest-NextBestTDsamples.R
+#'
+setMethod(
+  f = "nextBest",
+  signature = signature(
+    nextBest = "NextBestTDsamples",
+    doselimit = "numeric",
+    samples = "Samples",
+    model = "LogisticIndepBeta",
+    data = "Data"
+  ),
+  definition = function(nextBest, doselimit, samples, model, data, ...) {
+    doselimit <- ifelse(missing(doselimit) || length(doselimit) == 0, Inf, doselimit)
+
+    # Generate TDtarget during a trial and TDtarget at the end of trial samples.
+    target_in_trial_samples <- dose(x = nextBest@targetDuringTrial, model, samples)
+    target_trial_end_samples <- dose(x = nextBest@targetEndOfTrial, model, samples)
+
+    # Derive the prior/posterior estimates based on two above samples.
+    target_in_trial_est <- as.numeric(nextBest@derive(TDsamples = target_in_trial_samples))
+    target_trial_end_est <- as.numeric(nextBest@derive(TDsamples = target_trial_end_samples))
+
+    # Get eligible doses that do not exceed maximum possible dose.
+    is_dose_eligible <- data@doseGrid <= doselimit
+    doses_eligible <- data@doseGrid[is_dose_eligible]
+
+    # Find the next doses in the doseGrid. The next dose is the dose at level
+    # closest and below the target_in_trial_est (or target_trial_end_est, respectively).
+    # h_find_interval assumes that elements in doses_eligible are strictly increasing.
+    next_best_level <- h_find_interval(target_in_trial_est, doses_eligible)
+    next_best <- doses_eligible[next_best_level]
+
+    next_best_level1 <- h_find_interval(target_trial_end_est, doses_eligible)
+    next_best1 <- doses_eligible[next_best_level1]
+
+    # 95% credibility interval.
+    CITDEOT <- as.numeric(quantile(target_trial_end_samples, probs = c(0.025, 0.975)))
+    TDEOT_ratio <- CITDEOT[2] / CITDEOT[1]
+
+    # Build plot.
```

Comment:
the function is very long, suggest to consider factoring out plot in separate helper function

Reply:
Yap, I agree. This was actually my question when started to work on `nextBest` methods - https://github.com/Roche/crmPack/pull/273#issuecomment-1178861547

Reply:
Hm I would not do a separate file necessarily, just putting the helper function above the method could work. And better do it incrementally as you hit those long methods instead of doing it all together at once

Reply:
I see. Tx @danielinteractive. However, we already have `helpers_rules.R` file for this purpose, for example it contains already some helpers like that for `nextBest` methods? 

Also - since this new helper function which creates a plot is to be used only by this particular `nextBest` methods do you think we need to add another set of unit test just for this helper? or it is ok if we test it within `nextBest` as it is already?

Reply:
In any case we need to test the plot results, so you can go with the simpler way, either unit test for the plot helper or unit test for the method.